x-pack/plugins/security_solution/server/lib/detection_engine/signals/utils.ts

```diff
@@ -661,3 +663,356 @@ export const createTotalHitsFromSearchResult = ({
+            index: indexPattern,
+            format: 'json',
+          }
+        );
+        return { [indexPattern]: totalIndices.map((item) => item.index), ...(await acc) };
+      } catch (exc) {
+        logger.error(
+          buildRuleMessage(`[-] failed to cat indices: ${JSON.stringify(indices, null, 2)}`)
+        );
+        logger.error(buildRuleMessage(`Exception: ${exc}`));
+        return { ...(await acc) };
+      }
+    }, {} as Promise<Record<string, string[]>>);
+    return res;
+  } catch (exc) {
+    logger.error(
+      buildRuleMessage(`[-] failed to cat indices: ${JSON.stringify(indices, null, 2)}`)
+    );
+    logger.error(buildRuleMessage(`Exception: ${exc}`));
+    return Promise.resolve({});
+  }
+};
+
+export const findIndicesWithTimestampAndWithout = (
+  indexPatternRegEx: Record<string, RegExp>,
+  indexPatternIndices: Record<string, string[]>,
+  timestampsAndIndices: TimestampsAndIndices,
+  timestamp: string,
+  logger: Logger,
+  buildRuleMessage: BuildRuleMessage
+) => {
+  let failingIndexes: string[] = [];
+  let successIndexes: string[] = [];
+  for (const indexPattern of Object.keys(indexPatternRegEx)) {
+    const matchingIndexes = timestampsAndIndices[timestamp].filter((index) =>
+      indexPatternRegEx[indexPattern].test(index)
+    );
+    const [indexesWithTimestampField, indexesMissingTimestampField] = partition(
+      (index) => matchingIndexes.some((idx) => idx === index),
+      indexPatternIndices[indexPattern]
+    );
+
+    if (indexesMissingTimestampField.length > 0) {
+      logger.error(
+        buildRuleMessage(
+          `The following indices ${JSON.stringify(
+            indexesMissingTimestampField,
+            null,
+            2
+          )} matching matching index pattern ${JSON.stringify(
+            indexPattern,
+            null,
+            2
+          )} are missing required field ${timestamp}.`
+        )
+      );
+      failingIndexes = [...failingIndexes, ...indexesMissingTimestampField];
+    }
+    successIndexes = [...successIndexes, ...indexesWithTimestampField];
+  }
+  return [successIndexes, failingIndexes];
+};
+
+/**
+ * The basic gist is find all indices that have the given timestamp(s)
+ * If there is a timestamp override, ensure all indices for all index
+ * patterns have that timestamp override. If an index pattern does not,
+ * check if that index has an '@timestamp' field and default to that.
+ *
+ * If no timestamp override field is found, use the '@timestamp' field.
+ * If an index in a given index pattern is missing the '@timestamp' field
+ * set the status of the rule to be a partial failure
+ *
+ * If all index patterns are missing the '@timestamp' field and the override
+ * set the status to error.
+ * @param indices string[]
+ * @param timestamps string[]
+ * @param services AlertServices
+ * @param logger Logger
+ * @param buildRuleMessage
+ */
+export const timestampFieldCheck = async (
+  indices: string[],
+  timestamps: string[],
```

Comment:
Since we require `@timestamp` to be the default timestamp and only allow a single timestamp override, it would simplify the code to pass them in either as a tuple or expose the default timestamp as a shared constant and only pass in the timestamp override. That would eliminate [this logic](https://github.com/elastic/kibana/pull/83481/files#diff-333497786c31c4722a433be429554127c7eb40198062ef5004a057da5c0a23cfR847) and make it clear to readers that we aren't trying to support an arbitrary number of timestamps with this code.